### ISSUE TITLE
Cannot reconnect after disconnection from remote

### ISSUE THREAD

**Issue by pcurrivan** [microsoft/vscode-remote-release#4441]
- VSCode Version: 1.53.0-insider
- Local OS Version: Windows 10
- Remote OS Version: Linux
- Remote Extension/Connection Type: SSH

This is a repost of this issue, which has not been fixed: #1187

Steps to Reproduce:

1. Connect to a remote using SSH
2. Disconnect from network (not sure how to reproduce this)
3. Wait some period of time until the UI shows "Disconnected from SSH: hostname"

At this point if you do not notice that you are disconnected and try to edit files, you are not able to save them (error message: "Failed to save filename: Unable to write file... (Canceled: Canceled)").  If you click the "Disconnected from SSH" widget on the bottom-left, and click the action "Remote SSH: Connect Current Window to Host...", nothing happens.  The only way to reconnect is to close the window or select "Close Remote Connection", which will discard all of your edits.  So you are forced to save all your edits to separate files, close the connection, then re-open the connection and re-apply your changes.  Instead, there should be an option to reconnect.

**Comment by plwalsh** [comment]
Interestingly, this resulted from a [new feature](https://github.com/microsoft/vscode-docs/blob/master/remote-release-notes/v1_53.md#less-disruptive-disconnects) added in the most recent v1.53 release. While the disconnects are indeed now "less disruptive", they're also incredibly unnoticeable. I agree with the motivation for the feature -- to avoid causing your vscode window(s) from stealing OS focus upon disconnect, when the system dialogs would pop up asking you to "Reload window". However, I now have no idea when a remote connection has disconnected on me, until I try to save a file or type in the terminal. At which point I finally look down to the bottom left-hand corner and see "Disconnected from SSH".

I'm not entirely sure why the system dialog wasn't just changed to a vscode toast message (or whatever you call the bottom right-hand corner popups that usually need to be dismissed). In my opinion, that would be a great initial fix to this issue -- a vscode toast message informing you about the disconnect and providing you a button to "Reload window". And then a great follow-up to that would be to add a "Reconnect" command that shows up in the *Remote-SSH:* command list after clicking on the bottom left-hand corner *Remote Development* status bar widget.

TL;DR: A workaround is to open the command palette and select "reload window". But I don't think that's a sufficient solution to the issue.

**Comment by AnthonyParkssonos** [comment]
im stuck with this and now "Setting up SSH Host" hangs on reload sometimes. i'm pretty sure this is due to cpptools extension completely screwing up my system's memory but still this is unacceptable. i am having to hard reboot my system just to use it.

**Comment by pcurrivan** [comment]
@plwalsh Thanks!  "Reload window" seems to work.  Maybe it should be added to the actions that appear when you click the "Disconnected from SSH" widget.

**Comment by plwalsh** [comment]
@pcurrivan I definitely think it should be. But it should also be a little more apparent to the user that you've been disconnected, instead of being hidden in the corner of the status bar.

**Comment by umdstu** [comment]
Experiencing this as well since I came into work Monday (and every morning since). Confirmed that Reload Window is a temporary work around. I'll also mention that all Remote - SSH commands available from the bottom left disconnected banner do not work except for Close Connection. The other 3 do nothing at all. Additionally, in case it matters, I am using setting `"window.dialogStyle": "custom"`.

**Version: 1.53.0
Commit: 8490d3dde47c57ba65ec40dd192d014fd2113496
Date: 2021-02-03T15:56:04.185Z
Electron: 11.2.1
Chrome: 87.0.4280.141
Node.js: 12.18.3
V8: 8.7.220.31-electron.0
OS: Linux x64 4.14.214-160.339.amzn2.x86_64**

**Comment by MartinMajewski** [comment]
Experiencing a similar behavior. It seems that if my MacBook Pro goes to sleep, the connection is lost. This never happens when having an open ssh pipe in the Terminal, only with the Remote-SSH extension.

This wouldn't be a big issue if "Connect current window to host..." would work, but it doesn't. Nothing happens at all.
Choosing "Connect to host..." does not do anything either. Only "Disconnect from host" works, prompting me to save changes on files I have edited without noticing that I'm disconnected already.

Saving will fail because of obvious reasons. So I have to cope with it and start a completely new session with all the workspace synchronization overhead.

**Remote-SSH Plugin: v0.63.0
ersion: 1.53.1
Commit: 5d424b828ada08e1eb9f95d6cb41120234ef57c7
Date: 2021-02-08T23:35:15.302Z (2 days ago)
Electron: 11.2.1
Chrome: 87.0.4280.141
Node.js: 12.18.3
V8: 8.7.220.31-electron.0
OS: Darwin x64 20.3.0 (Big Sur 11.2.1)**

**Comment by galactica147** [comment]
hitting the same issue, this is definitely a regression from previous version. 
And now I'm unable to reconnect, but can only create a new window and connect from there.

**Comment by ssbarnea** [comment]
Mine hungs also without any hints on why:
```
[09:31:08.810] Log Level: 2
[09:31:08.812] remote-ssh@0.64.0
[09:31:08.812] darwin x64
[09:31:08.812] SSH Resolver called for "ssh-remote+leno", attempt 1
[09:31:08.813] "remote.SSH.useLocalServer": true
[09:31:08.813] "remote.SSH.sshPath": undefined
[09:31:08.813] "remote.SSH.sshConfigurationFile": undefined
[09:31:08.813] "remote.SSH.useFlock": true
[09:31:08.813] "remote.SSH.lockfilesInTmp": false
[09:31:08.814] "remote.SSH.localServerDownload": auto
[09:31:08.814] "remote.SSH.remoteServerListenOnSocket": false
[09:31:08.814] "remote.SSH.showLoginTerminal": false
[09:31:08.814] "remote.SSH.defaultExtensions": []
[09:31:08.814] SSH Resolver called for host: leno
[09:31:08.814] Setting up SSH remote "leno"
[09:31:08.816] Acquiring local install lock: /var/folders/3q/pc1jcyjj3qqbj6_x7n4kbtqw0000gn/T/vscode-remote-ssh-leno-install.lock
[09:31:08.825] Looking for existing server data file at /Users/ssbarnea/Library/Application Support/Code/User/globalStorage/ms-vscode-remote.remote-ssh/vscode-ssh-host-leno-622cb03f7e070a9670c94bae1a45d78d7181fbd4-0.64.0/data.json
[09:31:08.826] Found local server running: {"remoteListeningOn":{"port":39665},"osReleaseId":"fedora","arch":"x86_64","webUiAccessToken":"","sshAuthSock":"/tmp/ssh-mEhwE4mNuI/agent.2196062","tmpDir":"/run/user/1000","platform":"linux","connectionToken":"aa11a1a1-1a11-11a1-1a11-1a111a1aaa11","pid":75900,"ipcHandlePath":"/var/folders/3q/pc1jcyjj3qqbj6_x7n4kbtqw0000gn/T/vscode-ssh-askpass-e8f82cbb3ce1d7ab8db61bbe6191d97794f416b5.sock","socksPort":50778,"startupTime":1613467819187}
[09:31:08.833] Found running server - short-circuiting install
[09:31:08.834] Starting forwarding server. localPort 50828 -> socksPort 50778 -> remotePort 39665
```

**Comment by umdstu** [comment]
I don't think this is the same thing @ssbarnea 

But as an update in general for this issue, what I discovered this morning was that if you attempt to reload your windows (cmd + p -> Reload window) and you haven't yet reconnected (in my case, restarting the docker containers) your tunnel/proxy, all vscode windows close / the program completely exits. This really ought to be moved up in priority.

**Comment by MartinMajewski** [comment]
> I don't think this is the same thing @ssbarnea
> 
> But as an update in general for this issue, what I discovered this morning was that if you attempt to reload your windows (cmd + p -> Reload window) and you haven't yet reconnected (in my case, restarting the docker containers) your tunnel/proxy, all vscode windows close / the program completely exits. This really ought to be moved up in priority.

At least for a plain SSH-Remote setup reloading the entire window helps and is the least cumbersome process.

**Comment by umdstu** [comment]
> > I don't think this is the same thing @ssbarnea
> > But as an update in general for this issue, what I discovered this morning was that if you attempt to reload your windows (cmd + p -> Reload window) and you haven't yet reconnected (in my case, restarting the docker containers) your tunnel/proxy, all vscode windows close / the program completely exits. This really ought to be moved up in priority.
> 
> At least for a plain SSH-Remote setup reloading the entire window helps and is the least cumbersome process.

It definitely is, assuming your connection still is open. I was simply saying that if your connection is not open and you try to do as you suggested, vscode shuts down.

**Comment by MartinMajewski** [comment]
> > > I don't think this is the same thing @ssbarnea
> > > But as an update in general for this issue, what I discovered this morning was that if you attempt to reload your windows (cmd + p -> Reload window) and you haven't yet reconnected (in my case, restarting the docker containers) your tunnel/proxy, all vscode windows close / the program completely exits. This really ought to be moved up in priority.
> > 
> > 
> > At least for a plain SSH-Remote setup reloading the entire window helps and is the least cumbersome process.
> 
> It definitely is, assuming your connection still is open. I was simply saying that if your connection is not open and you try to do as you suggested, vscode shuts down.

For me this is not the case. When my connection to the Raspberry Pi is lost due to... reasons (?!)... reloading the VSC window reconnects immediately to the Pi and sets everything correct again.

**Comment by umdstu** [comment]
> > > > I don't think this is the same thing @ssbarnea
> > > > But as an update in general for this issue, what I discovered this morning was that if you attempt to reload your windows (cmd + p -> Reload window) and you haven't yet reconnected (in my case, restarting the docker containers) your tunnel/proxy, all vscode windows close / the program completely exits. This really ought to be moved up in priority.
> > > 
> > > 
> > > At least for a plain SSH-Remote setup reloading the entire window helps and is the least cumbersome process.
> > 
> > 
> > It definitely is, assuming your connection still is open. I was simply saying that if your connection is not open and you try to do as you suggested, vscode shuts down.
> 
> For me this is not the case. When my connection to the Raspberry Pi is lost due to... reasons (?!)... reloading the VSC window reconnects immediately to the Pi and sets everything correct again.

We are talking about two separate scenarios. Your connection may have been lost to the Pi, but your Pi is still accessible. However, if a Docker container is no longer spun up, it is no longer accessible. So reconnecting in VScode would, prior to this bug, have re-prompted to reconnect if it couldn't reach it, instead vscode shuts down if you try to reload and the host is no longer accessible.

**Comment by MartinMajewski** [comment]
> > > > > I don't think this is the same thing @ssbarnea
> > > > > But as an update in general for this issue, what I discovered this morning was that if you attempt to reload your windows (cmd + p -> Reload window) and you haven't yet reconnected (in my case, restarting the docker containers) your tunnel/proxy, all vscode windows close / the program completely exits. This really ought to be moved up in priority.
> > > > 
> > > > 
> > > > At least for a plain SSH-Remote setup reloading the entire window helps and is the least cumbersome process.
> > > 
> > > 
> > > It definitely is, assuming your connection still is open. I was simply saying that if your connection is not open and you try to do as you suggested, vscode shuts down.
> > 
> > 
> > For me this is not the case. When my connection to the Raspberry Pi is lost due to... reasons (?!)... reloading the VSC window reconnects immediately to the Pi and sets everything correct again.
> 
> We are talking about two separate scenarios. Your connection may have been lost to the Pi, but your Pi is still accessible. However, if a Docker container is no longer spun up, it is no longer accessible. So reconnecting in VScode would, prior to this bug, have re-prompted to reconnect if it couldn't reach it, instead vscode shuts down if you try to reload and the host is no longer accessible.

That's true. I just wanted to point out, that (like I said) in a plain ssh-remote scenario the reload workaround is a feasible way.

**Comment by umdstu** [comment]
> > > > > > I don't think this is the same thing @ssbarnea
> > > > > > But as an update in general for this issue, what I discovered this morning was that if you attempt to reload your windows (cmd + p -> Reload window) and you haven't yet reconnected (in my case, restarting the docker containers) your tunnel/proxy, all vscode windows close / the program completely exits. This really ought to be moved up in priority.
> > > > > 
> > > > > 
> > > > > At least for a plain SSH-Remote setup reloading the entire window helps and is the least cumbersome process.
> > > > 
> > > > 
> > > > It definitely is, assuming your connection still is open. I was simply saying that if your connection is not open and you try to do as you suggested, vscode shuts down.
> > > 
> > > 
> > > For me this is not the case. When my connection to the Raspberry Pi is lost due to... reasons (?!)... reloading the VSC window reconnects immediately to the Pi and sets everything correct again.
> > 
> > 
> > We are talking about two separate scenarios. Your connection may have been lost to the Pi, but your Pi is still accessible. However, if a Docker container is no longer spun up, it is no longer accessible. So reconnecting in VScode would, prior to this bug, have re-prompted to reconnect if it couldn't reach it, instead vscode shuts down if you try to reload and the host is no longer accessible.
> 
> That's true. I just wanted to point out, that (like I said) in a plain ssh-remote scenario the reload workaround is a feasible way.

I already agreed that that was a feasible workaround in my first comment to this issue :)

**Comment by afletch** [comment]
Also seeing this same problem when connection to the remote has been interrupted (in my case, laptop has been put to sleep). The workaround of command palette > reload window works, but something is obviously amiss here as reconnection should be automatic.

Version: 1.53.2
Commit: 622cb03f7e070a9670c94bae1a45d78d7181fbd4
Date: 2021-02-11T11:45:54.515Z
Electron: 11.2.1
Chrome: 87.0.4280.141
Node.js: 12.18.3
V8: 8.7.220.31-electron.0
OS: Darwin x64 19.6.0

**Comment by plwalsh** [comment]
The amount of times I've tried to click on a different editor tab, only for it to disappear with an error message saying it doesn't exist anymore because my remote-ssh has quietly disconnected is growing aggravatingly large.

I very much miss the blocking pop-up dialog telling me to reload the window because I've been disconnected.

**Comment by umdstu** [comment]
@roblourens Could we get confirmation that this is indeed a sort of regression with the appropriate labels added?

Thanks!

**Comment by roblourens** [comment]
This was fixed in https://github.com/microsoft/vscode-remote-release/issues/4389, in the nightly build currently, will be out in the next stable release. Sorry for not being on top of it earlier.

The thing mentioned in the release notes is actually not related. Successful reconnections should be less disruptive, but a real disconnection should always pop up a dialog.

**Comment by umdstu** [comment]
Awesome, thank you!
